### ISSUE TITLE
[TW#16441] BLE disconnects after 40 seconds

### ISSUE THREAD

**Issue by ayavilevich** [espressif/esp-idf#1262]
Hi,

I am using the latest esp-idf at this time. I build and run the gatt_server example.
I connect to the server with my app or "nRF Connect" app. The console shows a connect then a disconnect after 40 seconds. I connect again and the same thing happens.

This is with LG G4 phone Android version 6.0 . Couldn't reproduce with another device.

Any idea where the fault is? How to debug/troubleshoot?

```
Flashing binaries to serial port COM15 (app at offset 0x10000)...
esptool.py v2.1
Connecting........_
Chip is ESP32D0WDQ6 (revision 0)
Uploading stub...
Running stub...
Stub running...
Configuring flash size...
Auto-detected Flash size: 4MB
Flash params set to 0x0220
Compressed 19680 bytes to 11554...
Wrote 19680 bytes (11554 compressed) at 0x00001000 in 1.0 seconds (effective 152.8 kbit/s)...
Hash of data verified.
Compressed 692944 bytes to 374330...
Wrote 692944 bytes (374330 compressed) at 0x00010000 in 33.2 seconds (effective 166.8 kbit/s)...
Hash of data verified.
Compressed 3072 bytes to 82...
Wrote 3072 bytes (82 compressed) at 0x00008000 in 0.0 seconds (effective 1627.3 kbit/s)...
Hash of data verified.

Leaving...
Hard resetting...
MONITOR
--- idf_monitor on COM15 115200 ---
--- Quit: Ctrl+] | Menu: Ctrl+T | Help: Ctrl+T followed by Ctrl+H ---
ets Jun  8 2016 00:22:57

rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
ets Jun  8 2016 00:22:57

rst:0x10 (RTCWDT_RTC_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0018,len:4
load:0x3fff001c,len:5772
load:0x40078000,len:0
load:0x40078000,len:13804
entry 0x40079030
I (69) boot: Detected ESP32
I (32) boot: ESP-IDF v3.0-dev-1147-g8e04b861 2nd stage bootloader
I (32) boot: compile time 00:18:38
I (32) boot: Enabling RNG early entropy source...
I (38) boot: SPI Speed      : 40MHz
I (42) boot: SPI Mode       : DIO
I (46) boot: SPI Flash Size : 4MB
I (50) boot: Partition Table:
I (54) boot: ## Label            Usage          Type ST Offset   Length
I (61) boot:  0 nvs              WiFi data        01 02 00009000 00006000
I (69) boot:  1 phy_init         RF data          01 01 0000f000 00001000
I (76) boot:  2 factory          factory app      00 00 00010000 00100000
I (84) boot: End of partition table
I (88) esp_image: segment 0: paddr=0x00010020 vaddr=0x3f400020 size=0x2b4d4 (177
364) map
I (158) esp_image: segment 1: paddr=0x0003b4fc vaddr=0x3ffc0000 size=0x02b44 ( 1
1076) load
I (163) esp_image: segment 2: paddr=0x0003e048 vaddr=0x40080000 size=0x00400 (
1024) load
0x40080000: _iram_start at C:/prog/esp32/msys32/home/arik/esp/esp-idf/components
/freertos/xtensa_vectors.S:1685

I (166) esp_image: segment 3: paddr=0x0003e450 vaddr=0x40080400 size=0x01bc0 (
7104) load
I (177) esp_image: segment 4: paddr=0x00040018 vaddr=0x400d0018 size=0x68a9c (42
8700) map
0x400d0018: _stext at ??:?

I (332) esp_image: segment 5: paddr=0x000a8abc vaddr=0x40081fc0 size=0x107e8 ( 6
7560) load
0x40081fc0: _xt_user_exc at C:/prog/esp32/msys32/home/arik/esp/esp-idf/component
s/freertos/xtensa_vectors.S:651

I (360) esp_image: segment 6: paddr=0x000b92ac vaddr=0x400c0000 size=0x00000 (
   0) load
I (371) boot: Loaded app from partition at offset 0x10000
I (371) boot: Disabling RNG early entropy source...
I (372) cpu_start: Pro cpu up.
I (375) cpu_start: Starting app cpu, entry point is 0x40080eec
0x40080eec: call_start_cpu1 at C:/prog/esp32/msys32/home/arik/esp/esp-idf/compon
ents/esp32/cpu_start.c:222

I (0) cpu_start: App cpu up.
I (386) heap_init: Initializing. RAM available for dynamic allocation:
I (392) heap_init: At 3FFAFF10 len 000000F0 (0 KiB): DRAM
I (398) heap_init: At 3FFCA9F8 len 00015608 (85 KiB): DRAM
I (405) heap_init: At 3FFE0440 len 00003BC0 (14 KiB): D/IRAM
I (411) heap_init: At 3FFE4350 len 0001BCB0 (111 KiB): D/IRAM
I (417) heap_init: At 400927A8 len 0000D858 (54 KiB): IRAM
I (424) cpu_start: Pro cpu start user code
I (106) cpu_start: Starting scheduler on PRO CPU.
I (0) cpu_start: Starting scheduler on APP CPU.
I (173) system_api: Base MAC address is not set, read default base MAC address f
rom BLK0 of EFUSE
I (433) phy: phy_version: 366.0, ba9923d, Oct 31 2017, 18:06:17, 0, 0
I (473) GATTS_DEMO: REGISTER_APP_EVT, status 0, app_id 0

I (473) GATTS_DEMO: CREATE_SERVICE_EVT, status 0,  service_handle 40

I (473) GATTS_DEMO: SERVICE_START_EVT, status 0, service_handle 40

I (483) GATTS_DEMO: ADD_CHAR_EVT, status 0,  attr_handle 42, service_handle 40

I (493) GATTS_DEMO: the gatts demo char length = 3

I (493) GATTS_DEMO: prf_char[0] =11

I (503) GATTS_DEMO: prf_char[1] =22

I (503) GATTS_DEMO: prf_char[2] =33

I (503) GATTS_DEMO: ADD_DESCR_EVT, status 0, attr_handle 43, service_handle 40

I (513) GATTS_DEMO: REGISTER_APP_EVT, status 0, app_id 1

I (523) GATTS_DEMO: CREATE_SERVICE_EVT, status 0,  service_handle 44

I (533) GATTS_DEMO: SERVICE_START_EVT, status 0, service_handle 44

I (533) GATTS_DEMO: ADD_CHAR_EVT, status 0,  attr_handle 46, service_handle 44

I (543) GATTS_DEMO: ADD_DESCR_EVT, status 0, attr_handle 47, service_handle 44

I (38873) GATTS_DEMO: ESP_GATTS_CONNECT_EVT, conn_id 0, remote 41:67:d8:69:ef:ac
:
I (38873) GATTS_DEMO: CONNECT_EVT, conn_id 0, remote 41:67:d8:69:ef:ac:
I (39503) GATTS_DEMO: update connetion params status = 0, min_int = 16, max_int
= 32,conn_int = 6,latency = 0, timeout = 2000
I (79853) GATTS_DEMO: ESP_GATTS_DISCONNECT_EVT
I (134093) GATTS_DEMO: ESP_GATTS_CONNECT_EVT, conn_id 0, remote 41:67:d8:69:ef:a
c:
I (134103) GATTS_DEMO: CONNECT_EVT, conn_id 0, remote 41:67:d8:69:ef:ac:
I (134583) GATTS_DEMO: update connetion params status = 0, min_int = 16, max_int
 = 32,conn_int = 6,latency = 0, timeout = 2000
I (174923) GATTS_DEMO: ESP_GATTS_DISCONNECT_EVT
```

**Comment by ayavilevich** [comment]
Tried to reproduce the next day with the same setup and it didn't happen. However the console now shows a new "update connetion params status" line that was not there when I had the issue in the first place.

```
I (13548) GATTS_DEMO: ESP_GATTS_CONNECT_EVT, conn_id 0, remote 7c:60:38:ed:9d:56:
I (13548) GATTS_DEMO: CONNECT_EVT, conn_id 0, remote 7c:60:38:ed:9d:56:
I (14038) GATTS_DEMO: update connetion params status = 0, min_int = 16, max_int = 32,conn_int = 6,latency = 0, timeout = 2000
I (14578) GATTS_DEMO: update connetion params status = 0, min_int = 0, max_int = 0,conn_int = 24,latency = 0, timeout = 400
I (91168) GATTS_DEMO: ESP_GATTS_DISCONNECT_EVT
```

It seems that sometimes something fails in the process of updating connection parameters, which leads to disconnects down the line. When that happens it is consistent.

Hope it helps to pin point the issue.

**Comment by ayavilevich** [comment]
Made more tests and getting the "unstable" situation again. Perhaps not related to "connection parameters" as I am getting this issue with various "update connection" lines in the console.

```
I (512302) GATTS_DEMO: ESP_GATTS_CONNECT_EVT, conn_id 0, remote 77:e8:17:dd:0d:e2:
I (512302) GATTS_DEMO: CONNECT_EVT, conn_id 0, remote 77:e8:17:dd:0d:e2:
I (512622) GATTS_DEMO: update connetion params status = 0, min_int = 16, max_int = 32,conn_int = 2
4,latency = 0, timeout = 400
I (521152) GATTS_DEMO: ESP_GATTS_DISCONNECT_EVT
I (522292) GATTS_DEMO: ESP_GATTS_CONNECT_EVT, conn_id 0, remote 77:e8:17:dd:0d:e2:
I (522292) GATTS_DEMO: CONNECT_EVT, conn_id 0, remote 77:e8:17:dd:0d:e2:
I (523302) GATTS_DEMO: update connetion params status = 16, min_int = 16, max_int = 32,conn_int =
24,latency = 0, timeout = 2000
I (539862) GATTS_DEMO: ESP_GATTS_DISCONNECT_EVT
I (542252) GATTS_DEMO: ESP_GATTS_CONNECT_EVT, conn_id 0, remote 77:e8:17:dd:0d:e2:
I (542252) GATTS_DEMO: CONNECT_EVT, conn_id 0, remote 77:e8:17:dd:0d:e2:
I (543262) GATTS_DEMO: update connetion params status = 16, min_int = 16, max_int = 32,conn_int =
24,latency = 0, timeout = 2000
I (554252) GATTS_DEMO: ESP_GATTS_DISCONNECT_EVT
I (557312) GATTS_DEMO: ESP_GATTS_CONNECT_EVT, conn_id 0, remote 77:e8:17:dd:0d:e2:
I (557312) GATTS_DEMO: CONNECT_EVT, conn_id 0, remote 77:e8:17:dd:0d:e2:
I (558312) GATTS_DEMO: update connetion params status = 16, min_int = 16, max_int = 32,conn_int =
24,latency = 0, timeout = 2000
I (569312) GATTS_DEMO: ESP_GATTS_DISCONNECT_EVT
I (587272) GATTS_DEMO: ESP_GATTS_CONNECT_EVT, conn_id 0, remote 77:e8:17:dd:0d:e2:
I (587272) GATTS_DEMO: CONNECT_EVT, conn_id 0, remote 77:e8:17:dd:0d:e2:
I (587772) GATTS_DEMO: update connetion params status = 0, min_int = 16, max_int = 32,conn_int = 2
4,latency = 0, timeout = 400
I (596062) GATTS_DEMO: ESP_GATTS_DISCONNECT_EVT
I (597372) GATTS_DEMO: ESP_GATTS_CONNECT_EVT, conn_id 0, remote 77:e8:17:dd:0d:e2:
I (597382) GATTS_DEMO: CONNECT_EVT, conn_id 0, remote 77:e8:17:dd:0d:e2:
I (597882) GATTS_DEMO: update connetion params status = 0, min_int = 16, max_int = 32,conn_int = 2
4,latency = 0, timeout = 400
I (599172) GATTS_DEMO: ESP_GATTS_DISCONNECT_EVT
```

I suspect this is some corruption is BT state in the Android caused by specific flow which I will figure out. Once in this state, I need to restart BT on the phone to get back to properly working situation.

I will make more tests to see if this is specific to ESP32 and update.

**Comment by ayavilevich** [comment]
This one is a harder to reproduce invalid state that is caused by the same flow as #1263 .
I have included reproduction instructions in #1263 and will close this one as this is less important.
